SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduler.java

```diff
@@ -849,10 +849,19 @@ private boolean shouldRetryImmediately(SingularityRequest request, SingularityDe
       return false;
     }
 
-    if (task.isPresent()
-        && task.get().getTaskRequest().getPendingTask().getPendingTaskId().getPendingType() == PendingType.IMMEDIATE
-        && request.getRequestType() == RequestType.SCHEDULED) {
-      return false; // don't retry UI triggered scheduled jobs (UI triggered on-demand jobs are okay to retry though)
+    if (task.isPresent()) {
+      Optional<SingularityTaskHistoryUpdate> taskHistoryUpdate = taskManager.getTaskHistoryUpdate(task.get().getTaskId(), ExtendedTaskState.TASK_CLEANING);
+
+      if (taskHistoryUpdate.isPresent()
+          && task.get().getTaskRequest().getPendingTask().getPendingTaskId().getPendingType() == PendingType.ONEOFF
+          && taskHistoryUpdate.get().getStatusMessage().orElse("").contains("USER_REQUESTED")) {
+        return false; // don't retry one-off launches of on-demand jobs if they were killed by the user
+      }
+
+      if (task.get().getTaskRequest().getPendingTask().getPendingTaskId().getPendingType() == PendingType.IMMEDIATE
```

Comment:
small nit, can we do this check first for efficiency since it doesn't require the task history lookup?